File: backend/src/seadex_api.py
import requests


class SeadexApi:
    def get_release_data(self, anilist_id):
        base_url = "https://releases.moe/api/collections/entries/records"
        trs_url = f"{base_url}?filter=alID={anilist_id}&expand=trs"
        response = requests.get(trs_url)
        response.raise_for_status()

        data = response.json()

        items = data.get("items")
        items = items[0] if items else None

        comparison = items.get("comparison")
        notes = items.get("notes")
        theoretical_best = items.get("theoreticalBest")

        trs = items.get("expand")
        trs = trs.get("trs")

        for entry in trs:
            tracker = entry.get("tracker")
            release_group = entry.get("releaseGroup")
            url = entry.get("url")
            dual_audio = entry.get("dualAudio")
            is_best = entry.get("isBest")

            files = entry.get("files")
            total_file_size = 0

            episode_list = []

            for file in files:
                name = file.get("name")
                file_size_format = self.format_file_size(file.get("length"))
                total_file_size += file.get("length")

                episode_list.append({"name": name, "size": file_size_format})

            # print(episode_list)

            total_file_size_format = self.format_file_size(total_file_size)

            if entry.get("infoHash") == "<redacted>":
                private_tracker = True
            else:
                private_tracker = False

    def format_file_size(self, bytes):
        megabytes = bytes / (1024**2)
        gigabytes = bytes / (1024**3)

        if gigabytes >= 1:
            return f"{round(gigabytes, 1)}" + " GiB"
        else:
            return f"{round(megabytes, 1)}" + " MiB"


if __name__ == "__main__":
    SeadexApi().get_release_data(18897)

File: backend/seadex_api.ts

type FileEntry = {
    name: string;
    length: number;
};

type TRSEntry = {
    tracker: string;
    releaseGroup: string;
    url: string;
    dualAudio: boolean;
    isBest: boolean;
    infoHash: string;
    files: FileEntry[];
};

type ReleaseData = {
    comparison: string | null;
    notes: string | null;
    "theoretical best": string | null;
    releases: {
        tracker: string;
        "release group": string;
        url: string;
        "dual audio": boolean;
        "is best": boolean;
        "private tracker": boolean;
        "file size": string;
        "episode list": {
            name: string;
            size: string;
        }[];
    }[];
};

export class SeadexApi {
    private formatFileSize(bytes: number): string {
        const megabytes = bytes / (1024 ** 2);
        const gigabytes = bytes / (1024 ** 3);

        if (gigabytes >= 1) {
            return `${gigabytes.toFixed(1)} GiB`;
        } else {
            return `${megabytes.toFixed(1)} MiB`;
        }
    }

    async getReleaseData(anilistId: number): Promise<ReleaseData | null> {
        const baseUrl = "https://releases.moe/api/collections/entries/records";
        const trsUrl = `${baseUrl}?filter=alID=${anilistId}&expand=trs`;

        const response = await fetch(trsUrl);
        if (!response.ok) {
            throw new Error(`HTTP error! Status: ${response.status}`);
        }

        const data = await response.json();
        let items = data?.items?.[0] || null;
        if (!items) return null;

        const releaseDict: ReleaseData = {
            comparison: items.comparison || null,
            notes: items.notes || null,
            "theoretical best": items.theoreticalBest || null,
            releases: []
        };

        const trs = items.expand?.trs || [];
        for (const entry of trs as TRSEntry[]) {
            let totalFileSize = 0;
            const episodeList = entry.files.map(file => {
                totalFileSize += file.length || 0;
                return {
                    name: file.name,
                    size: this.formatFileSize(file.length)
                };
            });

            const totalFileSizeFormat = this.formatFileSize(totalFileSize);
            const privateTracker = entry.infoHash === "<redacted>";

            releaseDict.releases.push({
                tracker: entry.tracker,
                "release group": entry.releaseGroup,
                url: entry.url,
                "dual audio": entry.dualAudio,
                "is best": entry.isBest,
                "private tracker": privateTracker,
                "file size": totalFileSizeFormat,
                "episode list": episodeList
            });
        }

        return releaseDict;
    }
}

const api = new SeadexApi();
api.getReleaseData(18897)
    .then(data => {
        console.log(JSON.stringify(data, null, 2));
    })
    .catch(err => console.error(err));
